src/python/pants/core/util_rules/pants_environment.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+import os
+import re
+from dataclasses import dataclass
+from typing import Dict, Mapping, Optional, Sequence
+
+from pants.engine.rules import side_effecting
+from pants.util.frozendict import FrozenDict
+from pants.util.meta import frozen_after_init
+
+logger = logging.getLogger(__name__)
+
+
+@side_effecting
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class PantsEnvironment:
+    """PantsEnvironment is a representation of the environment variables the currently-executing
+    pants process was invoked with.
+
+    Since the exact contents of the environment are very specific to a given invocation of pants,
+    this type is marked as side-effecting, and can only be requested as an input to an uncacheable
+    rule. An uncacheable rule should accept this type as an input, and use `get_subset` to filter
+    the environment rules down to only the ones other rules care about as inputs, in order to avoid
+    a lot of cache invalidation if the environment changes spuriously.
+    """
+
+    name_value_re = re.compile(r"([A-Za-z_]\w*)=(.*)")
```

Comment:
Do these need to be typed as `ClassVar[re.Pattern]` so they aren't treated as instance fields by the dataclass machinery?

Alternatively we could stick them directly in the module itself.

Reply:
They do! Good catch.